packages/context/src/context.ts

```diff
@@ -197,22 +192,49 @@ export class Context {
    *
    * @param keyWithPath The binding key, optionally suffixed with a path to the
    *   (deeply) nested property to retrieve.
+   * @returns A promise of the bound value.
+   */
+  get<T>(keyWithPath: string): Promise<T>;
+
+  /**
+   * Get the value bound to the given key, optionally return a (deep) property
+   * of the bound value.
+   *
+   * @example
+   *
+   * ```ts
+   * // get "rest" property from the value bound to "config"
+   * // use "undefined" when not config was provided
+   * const config = await ctx.get('config#rest', {optional: true});
+   * ```
+   *
+   * @param keyWithPath The binding key, optionally suffixed with a path to the
+   *   (deeply) nested property to retrieve.
    * @param optionsOrSession Options or session for resolution. An instance of
    * `ResolutionSession` is accepted for backward compatibility.
-   * @returns A promise of the bound value.
+   * @returns A promise of the bound value, or a promise of undefined when
+   * the optional binding was not found.
    */
-  get(
+  get<T>(
     keyWithPath: string,
     optionsOrSession?: ResolutionOptions | ResolutionSession,
-  ): Promise<BoundValue> {
+  ): Promise<T | undefined>;
+
+  // Implementation
+  get<T>(
+    keyWithPath: string,
+    optionsOrSession?: ResolutionOptions | ResolutionSession,
+  ): Promise<T | undefined> {
     /* istanbul ignore if */
     if (debug.enabled) {
       debug('Resolving binding: %s', keyWithPath);
     }
     try {
-      return Promise.resolve(
-        this.getValueOrPromise(keyWithPath, optionsOrSession),
+      const valueOrPromiseLike = this.getValueOrPromise<T>(
+        keyWithPath,
+        optionsOrSession,
       );
+      return Promise.resolve<T | undefined>(valueOrPromiseLike);
```

Comment:
The implementation can be simplified using `async await`. For example:
```ts
  // Implementation
  async get<T>(
    keyWithPath: string,
    optionsOrSession?: ResolutionOptions | ResolutionSession,
  ): Promise<T | undefined> {
    /* istanbul ignore if */
    if (debug.enabled) {
      debug('Resolving binding: %s', keyWithPath);
    }
    return await this.getValueOrPromise<T>(
      keyWithPath,
      optionsOrSession,
    );
  }
```

Reply:
I think we are/were intentionally avoiding async functions in performance-critical code paths. This pull request is also focusing on fixing type definitions, not necessarily rewriting code from promise-like to async/await style.

Having said that, if there is a consensus that Context functions should be rewritten to async/await style as part of this pull request, then I don't mind too much.

Reply:
On the second thought, I think the actual reason for using `Promise.resolve` instead of `await` was that I though it's not allowed to `await` a non-promise value.

I looked up docs and it looks it's perfectly fine to await a value that's not a promise, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await:

> If the value of the _expression_ following the `await` operator is not a `Promise`, it's converted to a [resolved Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve).

In that light, I agree `ctx.get` should be rewritten to an async function and I'll make the changes ASAP.